### ISSUE TITLE
Unit enum decoding failure returns an atom instead of raising an exception

### ISSUE THREAD

**Issue by dlesl** [rusterlium/rustler#375]
Hi and thanks for this awesome library! I just wanted to say that I found it a bit surprising that the code generated by `rustler_codegen` returns `Err(rustler::Error::Atom("invalid_variant"))` for unit enums, but in other cases (for example a missing key in a map) it returns `Err(rustler::Error::RaiseTerm(...))`. Would you be open to changing this behaviour?

**Comment by evnu** [comment]
Yes, the error handling in `rustler_codegen` is somewhat inconsistent. What would you as a user rather expect, `RaiseTerm` or an atom?

Note that this would be a breaking change, so we need to be careful.

**Comment by dlesl** [comment]
I would expect `RaiseTerm`, because other `Decoder` implementations return `BadArg`. So for example if you were decoding a complex object like a map it would be surprising if a wrong type where a binary is expected raises `badarg`, but an incorrect atom results in the function returning `invalid_variant`. One option might be to implement both, as methods on the `Decoder` trait? I'm currently doing this:

```rust
pub trait DecodeOrRaise<'a>: Sized + 'a {
    fn decode_or_raise<T: Decoder<'a>>(self) -> NifResult<T>;
}

impl<'a> DecodeOrRaise<'a> for Term<'a> {
    fn decode_or_raise<T: Decoder<'a>>(self) -> NifResult<T> {
        use rustler::Error::*;
        match self.decode() {
            Ok(res) => Ok(res),
            Err(e) => Err(match e {
                BadArg => BadArg,
                Atom(s) | RaiseAtom(s) => RaiseAtom(s),
                Term(t) | RaiseTerm(t) => RaiseTerm(t),
            }),
        }
    }
}
```

**Comment by dlesl** [comment]
On second thoughts, there's no way to provide the "reverse" of `DecodeOrRaise`, since there's no non-raising version of `BadArg`, and since AFAICT `badarg` is the idiomatic way to handle e.g. a list where a binary was expected, I think the raising option is always right. If the user wants to handle the error in another way, like returning `{error, Reason}`, they can always convert the `Err` themselves (and IMO they would want to too, because there's no convention on what `Reason` should be).

**Comment by evnu** [comment]
@dlesl Sorry for taking so long to come back to you, I am quite swamped right now. I think it makes sense to change this behaviour and always use `RaiseTerm` in `rustler_codegen`. Would you be willing to implement that as a PR?
